File: Advanced Robotics Course/Chapter15/Lesson4/Chapter15_Lesson4.m
function [rmse, x_hist] = swarm_consensus_matlab(N, steps, p_fail, p_drop, eps, noise_std)
    if nargin < 1, N = 50; end
    if nargin < 2, steps = 200; end
    if nargin < 3, p_fail = 0.2; end
    if nargin < 4, p_drop = 0.1; end
    if nargin < 5, eps = 0.2; end
    if nargin < 6, noise_std = 0.01; end

    rng(0);
    x = -1 + 2 * rand(N, 1);
    alive = rand(N, 1) >= p_fail;

    x_hist = zeros(N, steps + 1);
    x_hist(:, 1) = x;

    for k = 1:steps
        x_new = x;
        for i = 1:N
            if ~alive(i), continue; end
            left = mod(i - 2, N) + 1;
            right = mod(i, N) + 1;

            sumDiff = 0;
            count = 0;

            if rand() >= p_drop && alive(left)
                sumDiff = sumDiff + (x(left) - x(i));
                count = count + 1;
            end
            if rand() >= p_drop && alive(right)
                sumDiff = sumDiff + (x(right) - x(i));
                count = count + 1;
            end

            if count > 0
                avgDiff = sumDiff / count;
                x_new(i) = x(i) + eps * avgDiff + noise_std * randn();
            end
        end
        x = x_new;
        x_hist(:, k + 1) = x;
    end

    idx_alive = find(alive);
    if isempty(idx_alive)
        rmse = 0;
        return;
    end
    mean_val = mean(x(idx_alive));
    rmse = sqrt(mean((x(idx_alive) - mean_val).^2));
end
      

File: Tutorials/Codes/Advanced_Robotics_Course/Chapter15/Lesson4/Chapter15_Lesson4.java
import java.util.Random;

public class SwarmConsensus {
    private final int N;
    private final int steps;
    private final double pFail;
    private final double pDrop;
    private final double eps;
    private final double noiseStd;
    private final double[] x;
    private final boolean[] alive;
    private final Random rng;

    public SwarmConsensus(int N, int steps,
                          double pFail, double pDrop,
                          double eps, double noiseStd) {
        this.N = N;
        this.steps = steps;
        this.pFail = pFail;
        this.pDrop = pDrop;
        this.eps = eps;
        this.noiseStd = noiseStd;
        this.x = new double[N];
        this.alive = new boolean[N];
        this.rng = new Random(0L);

        for (int i = 0; i < N; i++) {
            x[i] = -1.0 + 2.0 * rng.nextDouble();
            alive[i] = (rng.nextDouble() >= pFail);
        }
    }

    private double gauss() {
        // Box-Muller
        double u1 = rng.nextDouble();
        double u2 = rng.nextDouble();
        return Math.sqrt(-2.0 * Math.log(u1))
             * Math.cos(2.0 * Math.PI * u2) * noiseStd;
    }

    public void run() {
        for (int k = 0; k < steps; k++) {
            double[] xNew = x.clone();
            for (int i = 0; i < N; i++) {
                if (!alive[i]) continue;

                int left = (i - 1 + N) % N;
                int right = (i + 1) % N;

                double sumDiff = 0.0;
                int count = 0;

                if (rng.nextDouble() >= pDrop && alive[left]) {
                    sumDiff += x[left] - x[i];
                    count++;
                }
                if (rng.nextDouble() >= pDrop && alive[right]) {
                    sumDiff += x[right] - x[i];
                    count++;
                }

                if (count > 0) {
                    double avgDiff = sumDiff / (double) count;
                    xNew[i] = x[i] + eps * avgDiff + gauss();
                }
            }
            System.arraycopy(xNew, 0, x, 0, N);
        }
    }

    public double computeRMSE() {
        double sum = 0.0;
        int count = 0;
        for (int i = 0; i < N; i++) {
            if (alive[i]) {
                sum += x[i];
                count++;
            }
        }
        if (count == 0) return 0.0;
        double mean = sum / (double) count;

        double err2 = 0.0;
        for (int i = 0; i < N; i++) {
            if (alive[i]) {
                double d = x[i] - mean;
                err2 += d * d;
            }
        }
        return Math.sqrt(err2 / (double) count);
    }

    public static void main(String[] args) {
        SwarmConsensus sim = new SwarmConsensus(100, 300, 0.2, 0.1, 0.2, 0.01);
        sim.run();
        System.out.println("RMSE = " + sim.computeRMSE());
    }
}
      
